executor/common_linux.h

```diff
@@ -1485,6 +1485,218 @@ static long syz_io_uring_submit(volatile long a0, volatile long a1, volatile lon
+#define BTF_MAGIC 0xeB9F
+
+struct btf_header {
+	__u16 magic;
+	__u8 version;
+	__u8 flags;
+	__u32 hdr_len;
+	__u32 type_off;
+	__u32 type_len;
+	__u32 str_off;
+	__u32 str_len;
+};
+
+#define BTF_INFO_KIND(info) (((info) >> 24) & 0x0f)
+#define BTF_INFO_VLEN(info) ((info)&0xffff)
+
+#define BTF_KIND_INT 1
+#define BTF_KIND_ARRAY 3
+#define BTF_KIND_STRUCT 4
+#define BTF_KIND_UNION 5
+#define BTF_KIND_ENUM 6
+#define BTF_KIND_FUNC_PROTO 13
+#define BTF_KIND_VAR 14
+#define BTF_KIND_DATASEC 15
+
+struct btf_type {
+	__u32 name_off;
+	__u32 info;
+	union {
+		__u32 size;
+		__u32 type;
+	};
+};
+
+struct btf_enum {
+	__u32 name_off;
+	__s32 val;
+};
+
+struct btf_array {
+	__u32 type;
+	__u32 index_type;
+	__u32 nelems;
+};
+
+struct btf_member {
+	__u32 name_off;
+	__u32 type;
+	__u32 offset;
+};
+
+struct btf_param {
+	__u32 name_off;
+	__u32 type;
+};
+
+struct btf_var {
+	__u32 linkage;
+};
+
+struct btf_var_secinfo {
+	__u32 type;
+	__u32 offset;
+	__u32 size;
+};
+
+#define BTF_VMLINUX_FIX_ADDRESS ((char*)0x01000000)
+
+// Read out all the content of /sys/kernel/btf/vmlinux to the fixed address
+// buffer and return it. Return NULL if failed.
+static char* read_btf_vmlinux()
+{
+	int fd = open("/sys/kernel/btf/vmlinux", O_RDONLY);
+	if (fd < 0)
+		return NULL;
+
+	struct stat stat;
+	if (fstat(fd, &stat) < 0)
+		return NULL;
+	unsigned long vmlinux_size = stat.st_size;
+	// Read the whole vmlinux file into buf.
+	char* buf;
+	buf = (char*)mmap(BTF_VMLINUX_FIX_ADDRESS, vmlinux_size, PROT_READ | PROT_WRITE,
+			  MAP_SHARED | MAP_ANONYMOUS | MAP_FIXED_NOREPLACE, -1, 0);
+
+	if ((void*)buf == MAP_FAILED) {
+		if (errno == EEXIST) {
+			// The mmap already exists and we will assume this is the one that
+			// was made by the previous call to this function.
+			return BTF_VMLINUX_FIX_ADDRESS;
+		} else {
+			return NULL;
+		}
+	}
+
+	if (buf != BTF_VMLINUX_FIX_ADDRESS) {
+		fail("The address is different even though we issued "
+		     "MAP_FIXED_NOREPLACE");
+	}
+
+	unsigned long bytes_read = 0;
+	while (bytes_read != vmlinux_size) {
+		ssize_t ret = read(fd, buf + bytes_read, vmlinux_size - bytes_read);
+		if (ret <= 0) {
+			munmap(buf, vmlinux_size);
+			return NULL;
+		}
+		bytes_read += ret;
+	}
+	return buf;
+}
+
+// Lookup the name `target` in `vmlinux` and return its btf_id. Return -1 if
+// failed.
+static long btf_id_by_name(char* vmlinux, char* target)
+{
+	struct btf_header* btf_header = (struct btf_header*)vmlinux;
+	if (btf_header->magic != BTF_MAGIC)
+		return -1;
+	// These offsets are bytes relative to the end of the header.
+	char* btf_type_sec = vmlinux + btf_header->hdr_len + btf_header->type_off;
+	char* btf_str_sec = vmlinux + btf_header->hdr_len + btf_header->str_off;
+	// Scan through the btf type section, and find a type description that
+	// match the provided name.
+	char* cur = btf_type_sec;
+	// BTF index starts at 1.
+	long idx = 1;
+	while ((unsigned int)(cur - btf_type_sec) < btf_header->type_len) {
```

Comment:
Pointer subtraction and then casting to unsigned is a little gross.  Would it be cleaner to make cur an unsigned int referring to position in bytes within the btf type section, compare it to type_len here, and then btf_type = (struct btf_type*)(btf_type_sec+cur) ?